src/witness_generator/affine_expression.rs

```diff
@@ -114,8 +123,10 @@ impl AffineExpression {
         known_constraints: &impl BitConstraintSet,
     ) -> EvalResult {
         // Try to solve directly.
-        if let Ok(result) = self.solve() {
-            return Ok(result);
+        match self.solve() {
+            Ok(result) => return Ok(result),
+            Err(ConstraintUnsatisfiable(e)) => return Err(ConstraintUnsatisfiable(e)),
+            Err(_) => {}
```

Comment:
Why are other errors ignored here?

Reply:
`ConstraintUnsatisfiable` means that we have something like `1=2`, so it does not make sense to try the other way using bit constraints. All other errors just mean "I could not find a value, but I don't know if it is satisfiable or not". So we continue with the other method of using bit constraints.

I'm not sure - is there a way to make that distinction more explicit in Rust?

Reply:
What I was thinking of was changing `EvalResult` to `Result<EvalValue, EvalError>` with `EvalValue` an enum with the current Ok value and all recoverable errors, and have `EvalError` be only the errors we don't recover from. Is there such a clean distinction between recoverable and non recoverable errors?

Reply:
Yes, there is such a clean distinction. I mean there are of course cases where we just tentatively ask for a solution and then decide the other way when there is a conflict. But if a ConstraintUnsatisfiable or BitConstraintUnsatisfiable (I think the wording is different) reaches the top-level, this should lead to a termination of the program.